### DISCUSSION TITLE
Upload asset chunks concurrently

### DISCUSSION THREAD

**Discussion by ghost** [dfinity/sdk#2644] [Maintenance Tasks]
This was not possible through the bootstrap, for two reasons:
- the bootstrap only allowed one connection (though this still allowed multiple requests to execute within a single block)
- idle connections leaving agent in a state where all requests failed with a "channel closed" error after a http/408 timeout response

It's uncertain whether this will be possible post-bootstrap, but somehow, someday, we need to be able to upload more than one chunk per block.

Splitting up calls to `call_and_wait()` into many `call()` before waiting for the requests to finish would bypass the idle connection issue.

**Comment by viviveevee** [comment]
@smallstepman do you by chance already have this on your radar?

**Comment by ghost** [comment]
This has been implemented: https://github.com/dfinity/sdk/blob/master/src/canisters/frontend/ic-asset/src/batch_upload/plumbing.rs#L359
